src/PointEval/PointEvalHandler.jl

```diff
@@ -103,26 +105,26 @@ end
 function _check_isoparametric_boundaries(::Type{RefHypercube{dim}}, x_local::Vec{dim, T}) where {dim, T}
     tol = sqrt(eps(T))
     # All in the range [-1, 1]
-    return all(x -> abs(x) - 1 < tol, x_local)
+    return all(x -> abs(x) - 1 ≤ tol, x_local)
 end
 
 # check if point is inside a cell based on isoparametric coordinate
 function _check_isoparametric_boundaries(::Type{RefSimplex{dim}}, x_local::Vec{dim, T}) where {dim, T}
     tol = sqrt(eps(T))
     # Positive and below the plane 1 - ξx - ξy - ξz
-    return all(x -> x > -tol, x_local) && sum(x_local) - 1 < tol
+    return all(x -> x ≥ -tol, x_local) && sum(x_local) - 1 < tol
 end
 
 # See https://discourse.julialang.org/t/finding-the-value-of-a-field-at-a-spatial-location-in-juafem/38975/2
 # TODO: should we make iteration params optional keyword arguments?
-function find_local_coordinate(interpolation, cell_coordinates::Vector{V}, global_coordinate::V) where {dim, T, V <: Vec{dim, T}}
+function find_local_coordinate(interpolation::IP, cell_coordinates::Vector{V}, global_coordinate::V) where {dim, T, V <: Vec{dim, T}, ref_shape, IP <: Interpolation{ref_shape}}
     n_basefuncs = getnbasefunctions(interpolation)
     @assert length(cell_coordinates) == n_basefuncs
-    local_guess = zero(V)
+    local_guess = sample_random_point(ref_shape)
```

Comment:
I would vote for not having this random if it can be avoided, see https://github.com/Ferrite-FEM/Ferrite.jl/pull/818#discussion_r1359338948

Reply:
First, let me give you two arguments for using randomization here over a single fixed guess.

1. There is no easy heuristic for a single initial guess which guarantees convergence from the nearest, because there is no trivial relation between basin of attractions for roots of the nonlinear problem. For the heuristic given in your comment you can easily construct examples (which we likely also hit in practice) such that the newton will be in the wrong basin and cannot converge, which is one issue I try to fix here.
2. For a grid search we can guarantee convergence with a *fine enough* grid. However, the grid might be quite large for nonlinear problems.

Now the presented randomization strategy is a tradeoff, which will also converge with enough samples (because we will eventually hit the basin of attraction), however in 3D we will likely use less points than a grid search, provided we have a proper random number generator.

However, if you have a better strategy on how we can guarantee to start in the basin of attraction (at best with 1 sample) I will take it. :)


On the other side we should make this algorithm swappable anyways. I thought about NonlinearSolve.jl as a dependency for the nonlinear problem (via extension?) and an external function to pass here to provide the outer loop (i.e. initial guess + eventual looping). What do you think?

Reply:
Since it is a nonlinear multimodal optimization problem the approach should be tackle it as such. This is why I agree with Dennis' opinion that either a full multistart strategy with a grid of initial guesses or a random point multistart with less points should be used, since both is often use in such cases. Other than that, I would agree with any conventional method for this type of problem (trust region or the like)

Reply:
> For the heuristic given in your comment you can easily construct examples (which we likely also hit in practice) such that the newton will be in the wrong basin and cannot converge, which is one issue I try to fix here.

I guess I'm underestimating the complexity of this problem then - thanks for the clarification! If you happen to have an example around, I would be interested in visualizing it with a Q&D code I have that shows the residual field over a `QuadraticQuadrilateral`. When trying random quad distortions and points, I've so far only ended up with nice problems. 

So sure, having guaranteed convergence would be better then! An option that is potentially quite efficient would be to have a reasonable heuristic first guess, followed by a multistart algorithm based on random guesses in case that one doesn't work.

Reply:
I think you are right. The second derivative for the minimization problem $min_\xi||x(\xi)-x_g||^2$ should be just J, so for an intact grid we should end up with a convex setup due to $det(J)>0$, as you have written on slack?